File: advent-of-code/2025/day10-2/main.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# SPDX-FileCopyrightText: 2025 nanoseeds

import sys
from dataclasses import dataclass
from typing import List

from ortools.linear_solver import pywraplp


@dataclass
class InputCase:
    clicks: List[List[int]]  # 每个元素是一个点击的索引列表，如 (3) -> [3], (1,3) -> [1,3]
    joltage: List[int]  # 目标状态，如 {3,5,4,7} -> [3,5,4,7]

    @staticmethod
    def tostring(vec: List[int]) -> str:
        return "{" + ",".join(str(x) for x in vec) + "}"


def parse_num_list(s: str) -> List[int]:
    """
    解析形如 (3) 或 (1,3) 或 {3,5,4,7} 的字符串为整数列表
    """
    # 去除括号
    s = s.strip()
    s = s.replace("(", "").replace(")", "")
    s = s.replace("{", "").replace("}", "")
    if not s:
        return []
    return [int(x) for x in s.split(",")]


def read() -> List[InputCase]:
    """
    从标准输入读取数据，解析为 InputCase 列表
    输入格式示例:
    [.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
    """
    datas: List[InputCase] = []
    with open("./resource/large.data.in") as input:
        lines = input.readlines()
    for line in lines:
        line = line.strip()
        if not line:
            break
        parts = line.split()
        # 第一个是 light_str (如 [.##.])，跳过
        # 最后一个是 joltage (如 {3,5,4,7})
        # 中间的是 clicks (如 (3) (1,3) ...)

        order_list: List[List[int]] = []
        for part in parts[1:]:
            clicks = parse_num_list(part)
            order_list.append(clicks)

        joltage = order_list.pop()  # 最后一个是目标
        datas.append(InputCase(clicks=order_list, joltage=joltage))

    return datas


def process_case(case: InputCase) -> int:
    """
    使用 OR-Tools 求解整数线性规划问题

    问题建模:
    - 变量: x_i = 第 i 个按钮按的次数 (非负整数)
    - 约束: 对于每个计数器 j, sum(x_i * A[i][j]) = joltage[j]
      其中 A[i][j] = 1 如果按钮 i 影响计数器 j, 否则为 0
    - 目标: 最小化 sum(x_i) (总按键次数)
    """
    num_counters = len(case.joltage)
    num_buttons = len(case.clicks)

    # 构建系数矩阵 A: A[i][j] = 1 表示按钮 i 会增加计数器 j
    # clicks[i] 是按钮 i 影响的计数器索引列表
    A = [[0] * num_counters for _ in range(num_buttons)]
    for i, click_indices in enumerate(case.clicks):
        for j in click_indices:
            if j < num_counters:
                A[i][j] = 1

    # 创建 solver
    solver = pywraplp.Solver.CreateSolver("SCIP")
    if not solver:
        # 尝试使用 CBC 作为备选
        solver = pywraplp.Solver.CreateSolver("CBC")
    if not solver:
        print("Solver not available", file=sys.stderr)
        return 0x3F3F3F3F

    # 定义变量: x[i] = 按钮 i 的按压次数，非负整数
    # 上界设为所有目标值之和（最坏情况）
    upper_bound = sum(case.joltage) + 1
    x = [solver.IntVar(0, upper_bound, f"x_{i}") for i in range(num_buttons)]

    # 添加约束: 对于每个计数器 j，所有按钮对它的贡献之和等于目标值
    # sum(x[i] * A[i][j] for i in range(num_buttons)) == joltage[j]
    for j in range(num_counters):
        constraint_expr = sum(x[i] * A[i][j] for i in range(num_buttons))
        solver.Add(constraint_expr == case.joltage[j])

    # 目标函数: 最小化总按键次数
    solver.Minimize(sum(x))

    # 求解
    status = solver.Solve()

    if status == pywraplp.Solver.OPTIMAL:
        result = int(solver.Objective().Value())
        return result
    elif status == pywraplp.Solver.INFEASIBLE:
        print(f"No solution for {InputCase.tostring(case.joltage)}", file=sys.stderr)
        return 0x3F3F3F3F
    else:
        print(f"Solver status: {status}", file=sys.stderr)
        return 0x3F3F3F3F


def cal(datas: List[InputCase]) -> int:
    """
    主计算函数，遍历所有 case 并累加结果
    """
    result = 0
    for data in datas:
        min_result = process_case(data)
        if min_result != 0x3F3F3F3F:
            result += min_result
    return result


def output(data: int) -> None:
    print(data)


def main():
    input_data = read()
    output_data = cal(input_data)
    output(output_data)


if __name__ == "__main__":
    main()

File: advent-of-code/2025/day10-2/main.cpp
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 nanoseeds

#include <algorithm>
#include <ranges>
#include <tuple>
#include <vector>
#include <iostream>
#include <string>
#include <deque>
#include <unordered_map>
#include <format>
#include <cassert>
#include <complex>
#include <queue>
#include <set>

#ifdef ALGORITHM_TEST_MACRO
namespace day10_2 {
#endif
using std::cin;
using std::cout;
using std::tuple;
using std::vector;
using std::string;
using std::format;
static constexpr const char next_line{'\n'};

class input_case {
public:
    vector<vector<size_t> > clicks;
    vector<size_t> joltage;

    static string tostring(const auto &vec) {
        string result = "{";
        for (const auto &item: vec) {
            result += std::to_string(item) + ",";
        }
        if (!vec.empty()) {
            result.pop_back(); // Remove last comma
        }
        result += "}";
        return result;
    }
};

using input_type = std::vector<input_case>;
using output_type = int64_t;

inline input_type read();

output_type cal(const input_type &data);

void output(const output_type &data);

int main() {
    const auto input_data = read();
    const auto output_data = cal(input_data);
    output(output_data);
    return 0;
}

vector<size_t> num_list_to_vector(const auto ranges) {
    vector<size_t> clicks{};
    const auto node_str = ranges;
    const auto node_str_str = string(node_str.cbegin(), node_str.cend());
    auto orders = node_str | std::views::filter([](const char ch) {
        return ch != '(' && ch != ')'
               && ch != '{' && ch != '}';
    }) | std::views::split(',');
    for (auto num_str: orders) {
        const auto num = std::stoll(std::string(num_str.cbegin(), num_str.cend()));
        clicks.push_back(num);
    }
    return clicks;
}

inline input_type read() {
    input_type datas;
    for (std::string line; std::getline(std::cin, line);) {
        if (line.empty()) {
            break;
        }
        auto ranges = std::views::split(line, ' ');
        auto iter = ranges.cbegin();
        const auto light_str = *iter;
        std::vector<vector<size_t> > order_list;
        for (++iter; iter != ranges.cend(); ++iter) {
            const vector<size_t> clicks = num_list_to_vector(*iter);
            order_list.push_back(clicks);
        }
        const auto joltage = order_list.back();
        order_list.pop_back();
        datas.emplace_back(order_list, joltage);
    }
    return datas;
}

struct vectors {
    vector<size_t> click;
    size_t multi;
};

// input_state 添加上node之后的状态
// 注意, 将输入的(3) 处理为 [0,0,0,1]这种状态
vector<size_t> add(const vector<size_t> &input_state, const vectors &nodes) {
    assert(input_state.size() == nodes.click.size());
    vector result{input_state};
    for (size_t i = 0; i < result.size(); i++) {
        result[i] += (nodes.click[i] * nodes.multi);
    }
    return result;
}

vector<size_t> sub(const vector<size_t> &input_state, const vectors &nodes) {
    assert(input_state.size() == nodes.click.size());
    vector result{input_state};
    for (size_t i = 0; i < result.size(); i++) {
        result[i] -= (nodes.click[i] * nodes.multi);
    }
    return result;
}

bool overweidht(const vector<size_t> &target, const vectors &check) {
    assert(target.size() == check.click.size());
    const auto size = target.size();
    for (size_t i = 0; i < size; i++) {
        if (target[i] < check.click[i]) {
            return false;
        }
    }
    return true;
}


// 输入一个std::span, 输出一个集合, 其中key为状态, value为到达状态的node列表
std::unordered_map<string, std::tuple<vector<size_t>, size_t> > visit(std::span<const vectors> clicks,
                                                                      const vector<size_t> &target) {
    const vector<size_t> init_state(target.size(), 0);
    std::unordered_map<string, std::tuple<vector<size_t>, size_t> > nodes_map;
    const auto init_state_string = input_case::tostring(init_state);
    nodes_map.emplace(init_state_string, std::make_tuple(init_state, 0));
    size_t order{0};
    for (const auto &click: clicks) {
        std::unordered_map<string, std::tuple<vector<size_t>, size_t> > temp_nodes_map;
        // 不添加的就是set里的原装
        for (const auto &[state_string, nodes]: nodes_map) {
            const auto &[state, nodes_size] = nodes;
            const auto new_state = add(state, click);
            if (!overweidht(target, click)) {
                // 组合已经越界, 可以忽略
                continue;
            }
            const auto new_size = nodes_size + click.multi;
            const auto new_state_string = input_case::tostring(new_state);
            if (!temp_nodes_map.contains(new_state_string)) {
                temp_nodes_map[new_state_string] = std::make_tuple(new_state, new_size);
            } else {
                const auto &[old_state, old_size] = temp_nodes_map[new_state_string];
                if (new_size < old_size) {
                    temp_nodes_map[new_state_string] = std::make_tuple(new_state, new_size);
                }
            }
        }
        for (const auto &[state, nodes]: temp_nodes_map) {
            if (!nodes_map.contains(state)) {
                nodes_map[state] = nodes;
            } else {
                const auto &[old_nodes, old_size] = nodes_map.at(state);
                const auto &[new_nodes, new_size] = nodes;
                if (new_size < old_size) {
                    nodes_map[state] = nodes;
                }
            }
        }
        order++;
    }
      return nodes_map;
}


// {3,5,4,7}最多可以由 7个 {0,0,0,1}, 可以分拆成 {1,2,4}
vector<vectors> extend(const vector<size_t> &joltage, const vector<size_t> &click) {
    size_t max_combine{0x3f3f3f3f};
    assert(joltage.size() == click.size());
    for (size_t i = 0; i < joltage.size(); i++) {
        if (click[i] != 0) {
            max_combine = std::min(max_combine, joltage[i] / click[i]);
        }
    }
    vector<size_t> multis{};
    for (size_t base = 1; max_combine > base;) {
        multis.push_back(base);
        max_combine -= base;
        base *= 2;
    }
    if (max_combine > 0) {
        multis.push_back(max_combine);
    }
    vector<vectors> result{};
    for (const auto multi: multis) {
        result.emplace_back(click, multi);
    }
    return result;
}

vector<size_t> points_to_vector(std::span<const size_t> points_span, size_t length) {
    vector<size_t> result(length, 0);
    for (const auto point: points_span) {
        result[point] += 1;
    }
    return result;
}

output_type cal(const input_type &datas) {
    // 按顺序分割成两部分, 分别求可能性, 再两两组合
    int64_t result{0};
    for (const auto &data: datas) {
        size_t min_result{0x3f3f3f3f};
        vector<vectors> clicks_extend{};
        for (const auto &click: data.clicks) {
            const vector<size_t> click_vector = points_to_vector(click, data.joltage.size());
            const auto extended = extend(data.joltage, click_vector);
            clicks_extend.insert(clicks_extend.end(), extended.cbegin(), extended.cend());
        }

        const auto nodes = clicks_extend.size();
        const auto half_nodes = nodes / 2;
        const auto left = std::span(clicks_extend.cbegin(), clicks_extend.cbegin() + half_nodes);
        const auto right = std::span(clicks_extend.cbegin() + half_nodes, clicks_extend.cend());

        const auto &target = data.joltage;

        const auto left_state = visit(left, target);
        const auto right_state = visit(right, target);
        const auto target_string = input_case::tostring(target);
        if (left_state.contains(target_string)) {
            const auto &[state, size] = left_state.at(target_string);
            min_result = size;
        }
        if (right_state.contains(target_string)) {
            const auto &[state, size] = right_state.at(target_string);
            min_result = std::min(min_result, size);
        }
        // 都没有直接相加的情况
        for (const auto &[left_key, value]: left_state) {
            const auto &[node, size] = value;
            const auto right_vec = sub(data.joltage, vectors{node, 1});
            const auto right_str = input_case::tostring(right_vec);
            if (right_state.contains(right_str)) {
                const auto &[right_node, right_size] = right_state.at(right_str);
                min_result = std::min(min_result, size + right_size);
            }
        }
        if (result != 0x3f3f3f3f) {
            result += min_result;
        }
    }
    return result;
}


void output(const output_type &data) {
    std::cout << data << next_line;
}

static const auto faster_streams = [] {
    srand(time(nullptr));
    // use time to init the random seed
    std::ios::sync_with_stdio(false);
    std::istream::sync_with_stdio(false);
    std::ostream::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    // 关闭c++风格输入输出 , 与C风格输入输出的同步,提高性能.
    return 0;
}();

#ifdef ALGORITHM_TEST_MACRO
}
#endif
